### ISSUE TITLE
Detect wrong argument constraint type at runtime

### ISSUE THREAD

**Issue by thomaslevesque** [FakeItEasy/FakeItEasy#1237]
Related to #1177, #1230, #1163 

We already have the FakeItEasy0005 diagnostic to detect this, but we can also detect it at runtime, and we should throw in this case.

A possible approach would be to examine all the constraints collected by `IArgumentConstraintTrapper.TrapConstraints` and check if their type is compatible with the parameter type.

**Comment by blairconrad** [comment]
I like the idea of simply rejecting all expressions that are conversions in `CheckArgumentExpressionIsValid`.
Sadly, we already allow the user to pass a `T` constraint in the place of a `Nullable<T>` parameter.

@thomaslevesque may be right: we might need to investigate the types that we get from the constraint trap.
Of course, we could just do this when the expression is a conversion. I'll play a bit.

**Comment by thomaslevesque** [comment]
From #1230:

> > True. I have started working on #1237 
> 
> Oh. So have I. It was unowned, so I self-assigned and started working on it last night.

As you like @blairconrad. I'm mostly done, but I haven't written any tests yet, and I won't have time to work on it for a few days. If you want to look at what I have, here it is: https://github.com/thomaslevesque/FakeItEasy/commit/41e601191e4d36ea8b3f22149001b0f6fdc0d5c0

I added a new `Type` property to `IArgumentConstraint` in order to be able to compare against the actual parameter type.

**Comment by blairconrad** [comment]
Oh, man. My approach was going to be _totally_ different. Which probably means I'm forgetting lots of stuff! I don't mind just noodling along as a hobby even if/when we take your fix.

I'm going to be away tomorrow and then Friday through Sunday as well, so who knows when I'll be done…

**Comment by thomaslevesque** [comment]
> My approach was going to be totally different

I look forward to seeing what you had in mind ;)

**Comment by blairconrad** [comment]
Well, here it is. Very rough. Tests at least need review and the production code needs refactor.
https://github.com/blairconrad/FakeItEasy/tree/check-type

**Comment by thomaslevesque** [comment]
It's an interesting approach. I like the fact that it doesn't require as much changes as mine. On the other hand, it doesn't detect wrong types in  hidden nested constraints, e.g. constraints created by user-defined methods.

**Comment by blairconrad** [comment]
Can you gimme an example of such a thing?

**Comment by thomaslevesque** [comment]
Something like this:

```
public static class Extensions
{
    public static int IsEven(this IArgumentConstraintManager<int> manager)
    {
        return manager.Matches(x => x % 2 == 0);
    }
}
```

Or even:

```
private static int AnEvenNumber()
{
    return A<int>.That.Matches(x => x % 2 == 0);
}
```

**Comment by blairconrad** [comment]
Ugh. You're right, of course. I need to look at your solution more closely to see how it works. I have to relearn argument constraint trapping every single time I look at it.

**Comment by blairconrad** [comment]
I found my refresher to be useful anyhow. Your approach has more promise. Abandoning mine.

**Comment by blairconrad** [comment]
This change has been released as part of [FakeItEasy 4.2.0](https://github.com/FakeItEasy/FakeItEasy/releases/tag/4.2.0).
